### ISSUE TITLE
Attributes intermittently not being saved in order for variable products

### ISSUE THREAD

**Issue by NickGreen** [woocommerce/woocommerce#19740]
## Prerequisites

- [x] I have searched for similar issues in both open and closed tickets and cannot find a duplicate
- [x] The issue still exists against the latest `master` branch of WooCommerce on Github (this is **not** the same version as on WordPress.org!)
- [x] I have attempted to find the simplest possible steps to reproduce the issue
- [ ] I have included a failing test as a pull request (Optional)

## Synopsis
When purchasing a variable product, the attributes aren't being saved in the order itemmeta, and therefore not displaying in the order. Example:
![](https://cldup.com/fs8QyPtu6H-3000x3000.png)

This doesn't happen every time, and here's an example of the attributes saving (and being displayed in the order):
![](https://cldup.com/EDhe3VnaE4-3000x3000.png)

We've verified that the attributes for these orders are not stored in `wp_woocommerce_order_itemmeta` table, so it's not a display issue.

This originated from a support ticket of ours, and we've been able to duplicate on the customer site as well as locally.

## Steps to reproduce the issue

1. Create a variable product with two attributes, one with 3 and one with 2, for example: https://cloudup.com/cHEnTlcai3t
2. Run through the checkout process
3. View the order, and see that the attributes aren't displayed in the line item meta.

## Expected/actual behavior

When I follow those steps, I see that the attributes aren't showing in the order line item meta, or in the database `wp_woocommerce_order_itemmeta` table. This is reproduceable, but happens maybe 60% of the time.

I was expecting to see the attributes saved and displayed every time.

## Isolating the problem

- [x] This bug happens with only WooCommerce plugin active
- [x] This bug happens with a default WordPress theme active, or [Storefront](https://woocommerce.com/storefront/)
- [x] I can reproduce this bug consistently using the steps above

## WordPress Environment

<details>
```


### WordPress Environment ###

Home URL: http://localhost/subscriptions-test-3
Site URL: http://localhost/subscriptions-test-3
WC Version: 3.3.0
Log Directory Writable: ✔
WP Version: 4.9.5
WP Multisite: –
WP Memory Limit: 256 MB
WP Debug Mode: ✔
WP Cron: ✔
Language: en_US
External object cache: –

### Server Environment ###

Server Info: Apache/2.2.31 (Unix) mod_wsgi/3.5 Python/2.7.13 PHP/7.1.1 mod_ssl/2.2.31 OpenSSL/1.0.2j DAV/2 mod_fastcgi/2.4.6 mod_perl/2.0.9 Perl/v5.24.0
PHP Version: ❌ 7.1.1 - We recommend using PHP version 7.2 or above for greater performance and security. How to update your PHP version
PHP Post Max Size: 32 MB
PHP Time Limit: 30
PHP Max Input Vars: 1000
cURL Version: 7.52.1
OpenSSL/1.0.2j

SUHOSIN Installed: –
MySQL Version: 5.6.35
Max Upload Size: 32 MB
Default Timezone is UTC: ✔
fsockopen/cURL: ✔
SoapClient: ✔
DOMDocument: ✔
GZip: ✔
Multibyte String: ✔
Remote Post: ✔
Remote Get: ✔

### Database ###

WC Database Version: 3.3.4
WC Database Prefix: wp_
MaxMind GeoIP Database: ❌ The MaxMind GeoIP Database does not exist - Geolocation will not function. You can download and install it manually from https://dev.maxmind.com/geoip/geoip2/geolite2/ to the path: . Scroll down to "Downloads" and download the "MaxMind DB binary
gzipped" file next to "GeoLite2 Country". Please remember to uncompress GeoLite2-Country_xxxxxxxx.tar.gz and upload the GeoLite2-Country.mmdb file only.

Total Database Size: 4.01MB
Database Data Size: 2.50MB
Database Index Size: 1.51MB
wp_woocommerce_sessions: Data: 0.02MB + Index: 0.02MB
wp_woocommerce_api_keys: Data: 0.02MB + Index: 0.03MB
wp_woocommerce_attribute_taxonomies: Data: 0.02MB + Index: 0.02MB
wp_woocommerce_downloadable_product_permissions: Data: 0.02MB + Index: 0.05MB
wp_woocommerce_order_items: Data: 0.02MB + Index: 0.02MB
wp_woocommerce_order_itemmeta: Data: 0.06MB + Index: 0.03MB
wp_woocommerce_tax_rates: Data: 0.02MB + Index: 0.06MB
wp_woocommerce_tax_rate_locations: Data: 0.02MB + Index: 0.03MB
wp_woocommerce_shipping_zones: Data: 0.02MB + Index: 0.00MB
wp_woocommerce_shipping_zone_locations: Data: 0.02MB + Index: 0.03MB
wp_woocommerce_shipping_zone_methods: Data: 0.02MB + Index: 0.00MB
wp_woocommerce_payment_tokens: Data: 0.02MB + Index: 0.02MB
wp_woocommerce_payment_tokenmeta: Data: 0.02MB + Index: 0.03MB
wp_woocommerce_log: Data: 0.02MB + Index: 0.02MB
wp_ariadminer_connections: Data: 0.02MB + Index: 0.00MB
wp_commentmeta: Data: 0.02MB + Index: 0.03MB
wp_comments: Data: 0.06MB + Index: 0.09MB
wp_links: Data: 0.02MB + Index: 0.02MB
wp_options: Data: 1.05MB + Index: 0.05MB
wp_postmeta: Data: 0.36MB + Index: 0.44MB
wp_posts: Data: 0.06MB + Index: 0.06MB
wp_rg_form: Data: 0.02MB + Index: 0.00MB
wp_rg_form_meta: Data: 0.02MB + Index: 0.00MB
wp_rg_form_view: Data: 0.02MB + Index: 0.03MB
wp_rg_incomplete_submissions: Data: 0.02MB + Index: 0.02MB
wp_rg_lead: Data: 0.02MB + Index: 0.03MB
wp_rg_lead_detail: Data: 0.02MB + Index: 0.06MB
wp_rg_lead_detail_long: Data: 0.02MB + Index: 0.00MB
wp_rg_lead_meta: Data: 0.02MB + Index: 0.05MB
wp_rg_lead_notes: Data: 0.02MB + Index: 0.03MB
wp_termmeta: Data: 0.02MB + Index: 0.03MB
wp_terms: Data: 0.02MB + Index: 0.03MB
wp_term_relationships: Data: 0.02MB + Index: 0.02MB
wp_term_taxonomy: Data: 0.02MB + Index: 0.03MB
wp_usermeta: Data: 0.02MB + Index: 0.03MB
wp_users: Data: 0.02MB + Index: 0.05MB
wp_wc_download_log: Data: 0.02MB + Index: 0.03MB
wp_wc_webhooks: Data: 0.02MB + Index: 0.02MB
wp_wpml_mails: Data: 0.25MB + Index: 0.00MB

### Post Type Counts ###

page: 7
payment_retry: 5
post: 2
product: 17
product_variation: 55
revision: 1
scheduled-action: 21
shop_coupon: 3
shop_order: 28
shop_subscription: 13
wc_membership_plan: 5
wc_user_membership: 2
wc_zapier_feed: 1

### Security ###

Secure connection (HTTPS): ❌
					Your store is not using HTTPS. Learn more about HTTPS and SSL Certificates.
Hide errors from visitors: ✔

### Active Plugins (2) ###

WooCommerce Stripe Gateway: by WooCommerce – 4.0.7
WooCommerce: by Automattic – 3.4.0-dev

### Settings ###

API Enabled: ✔
Force SSL: –
Currency: USD ($)
Currency Position: left
Thousand Separator: ,
Decimal Separator: .
Number of Decimals: 2
Taxonomies: Product Types: external (external)
grouped (grouped)
simple (simple)
subscription (subscription)
variable (variable)
variable subscription (variable-subscription)

Taxonomies: Product Visibility: exclude-from-catalog (exclude-from-catalog)
exclude-from-search (exclude-from-search)
featured (featured)
outofstock (outofstock)
rated-1 (rated-1)
rated-2 (rated-2)
rated-3 (rated-3)
rated-4 (rated-4)
rated-5 (rated-5)


### WC Pages ###

Shop base: #4 - /shop/
Cart: #5 - /cart/
Checkout: #6 - /checkout/
My account: #7 - /my-account/
Terms and conditions: ❌ Page not set

### Theme ###

Name: Twenty Seventeen
Version: 1.4 – 1.5 is available
Author URL: https://wordpress.org/
Child Theme: ❌ – If you are modifying WooCommerce on a parent theme that you did not build personally we recommend using a child theme. See: How to create a child theme
WooCommerce Support: ✔

### Templates ###

Overrides: –


```
</details>

**Comment by WPprodigy** [comment]
I will say I've seen this issue in the wild before.

**Comment by mikejolley** [comment]
Not sure what we can do with this report alone. I also cannot. This is where the data is passed to the order item:

https://github.com/woocommerce/woocommerce/blob/e8cc1076ad6d6f1d3932fb8087689e79f776768d/includes/class-wc-checkout.php#L373

This is then added as meta data on the object:

https://github.com/woocommerce/woocommerce/blob/bc7846c41205855faa4cbfde059df18ac07e33f0/includes/class-wc-order-item-product.php#L175

These are then stored in `woocommerce_order_itemmeta` on save and linked to the line item - note the line item will have a different ID from the order so be sure you're looking for the correct rows.

I point that out because it can also be display related; if the attribute forms part of the variation name it will be hidden. So for example in your first screenshot there is a 2, 1. These are the attributes you had at that point. So if those values exist as meta, they will be hidden from display. Perhaps test that it's not a display issue using this code:

`
add_action( 'woocommerce_is_attribute_in_product_name', '__reutrn_false' );
`

**Comment by alfonsberger** [comment]
We have this issue with some of our orders as well. Started right around end of March, so I guess this was from one of the latest updates?

**Comment by mikejolley** [comment]
More likely it's what I said above or we would have seen more complaints.

@NickGreen Do you have anything to add based on my comment above? We won't be able to do anything with this ticket unless it's properly reproducible.

**Comment by NickGreen** [comment]
@mikejolley I appreciate your explanation and trying to duplicate. I think there may have been a mixup on my part regarding where I saw the data not being saved. 

First of all, I added your suggested filter, and I'm able to see the meta displayed in both places:
![](https://cldup.com/8U5N5fXju2-3000x3000.png)

I then added the snippet to the customer's staging site, where I originally saw the data not being saved, and they still aren't showing up in either place:
![](https://cldup.com/pBHv7eHXk5-3000x3000.png)

This leads me to believe that it's a site-specific issue (since there's no hard proof otherwise), so I suppose this can be closed, and I'll ask them to open a ticket through Woo Support.
